### PR TITLE
Generic mcp2301x driver

### PR DESCRIPTION
**Pull request by rtitmuss**
## Description

Added a generic mcp2301x driver. This is based on the implementation from the Torn keyboard.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by rtitmuss** [comment]
Proposal as discussed in https://github.com/qmk/qmk_firmware/pull/13773

**Comment by tzarc** [comment]
Would need to retarget the `develop` branch, and also probably be situated in `drivers/gpio/xxxxx`.

**Comment by zvecr** [comment]
Along with the `drivers/gpio/` I also have a few general comments:
* Having matrix and encoder feature logic within drivers is bleeding too much of the implementation into the wrong location
  * The current abstraction isnt great for swapping out a portion of the logic, but right now i would expect something `CUSTOM_MATRIX = MCP2301x` with the define name maybe being reworked
* The implementation of the `MCP2301x_ENABLE` logic assumes too much of the hardware implementation and makes it difficult to use regular encoder logic with a mcp based matrix
  * I would expect some work to abstract out the backend of the encoder logic so it can be swapped out with a similar "_DRIVER" option per some of the other features

**Comment by rtitmuss** [comment]
Thanks for the comments. I have made some updates, please consider this PR as draft for further feedback. This includes core and keyboard changes for my sanity moving code around - I can separate these if required once we've agreed on this in principle.

- the mcp2301x gpio code has been moved to drivers/gpio.
- matrix scanning has been added as a CUSTOM_MATRIX option
- support for the encoder is kept as part of the keyboard definition for now (I don't think other keyboards use mcp2301x to support an encoder so there is little value in making this generic right now)

Note documentation needs updating, which I'll do have feedback. Thanks.

File: drivers/gpio/mcp2301x.c [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline drivers/gpio/mcp2301x.c:19]:
```suggestion
```

**Review by drashna — COMMENTED** [review]
One issue I have with this, is that there is no error handling for when the expander is disconnected? 

This means that if the half isn't connected, that it will cause serious performance issues until it's reconnected. 

And since it has no error handling for that case, it can't re-run init code on reconnect.

which is to say, that I do want support for this, but it means that the ZSA boards (the ergodox ez and moonlander) cannot and will not use this code as is.  which is to say, would ideally like to be able to support any board using this setup, but it's not parity with the custom code that already exists. 

And if that functionality is added, it needs a callback for re-connecting for the ZSA boards, too. 

Eg, this should ideally be a hybrid of the ergodox expander code and the torn expander code.

**Pull request closed without merging**